### PR TITLE
[storage/journal] implement contiguous::fixed journal using segmented::fixed

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Gets rid of some of the code duplication across them.

Resolves: https://github.com/commonwarexyz/monorepo/issues/2749

### CODE REVIEW

File: storage/src/journal/contiguous/fixed.rs [hunk 1]
```diff
@@ -449,85 +345,78 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+                                        items_per_blob - start_pos_in_section
+                                    } else {
+                                        items_per_blob
+                                    };
+
+                                    let got_items = if prev_section == start_section {
+                                        *last_pos - start_pos_in_section + 1
+                                    } else {
+                                        *last_pos + 1
+                                    };
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs]: The calculation is correct. If we used just *last_pos + 1, we'd incorrectly count positions we'd never read.

File: storage/src/journal/segmented/fixed.rs [hunk 2]
```diff
@@ -146,73 +146,118 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+                                        )),
                                     }
                                 }
+                                Err(RError::BlobInsufficientLength) => {
+                                    Some((
+                                        Err(Error::Corruption(format!(
+                                            "trailing bytes in section {section} at position {position}"
+                                        ))),
+                                        (section, buf, blob, reader, blob_size, valid_size, blob_size),
+                                    ))
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: I don't think that's correct, but since we're returning error which will abort the stream it doesn't matter so much anyway.

File: storage/src/journal/contiguous/fixed.rs [hunk 3]
```diff
@@ -413,32 +314,27 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
-            if offset >= self.tail.size().await {
-                return Err(Error::ItemOutOfRange(pos));
+        self.inner.get(section, pos_in_section).await.map_err(|e| {
+            // Map segmented errors to contiguous errors
+            match e {
+                Error::SectionOutOfRange(_)
+                | Error::AlreadyPrunedToSection(_)
+                | Error::ItemOutOfRange(_) => {
+                    unreachable!("range should have already been checked")
+                }
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:398]: getting rid of the gap detection is fine, but probably we should return error corrupted instead of panicking.

[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:398]: Changed to return Error::Corruption.

File: storage/src/journal/contiguous/fixed.rs [hunk 4]
```diff
@@ -413,185 +313,108 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+        // Check all non-tail sections in range are complete before starting the stream.
+        let oldest = self.inner.oldest_section().unwrap_or(start_section);
+        if let Some(newest) = self.inner.newest_section() {
+            for section in start_section.max(oldest)..newest {
+                let len = self.inner.section_len(section).await?;
+                let expected = if section == start_section {
+                    items_per_blob - start_pos_in_section
+                } else {
+                    items_per_blob
+                };
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs]: good bot, fixed.

File: storage/src/journal/segmented/fixed.rs [hunk 5]
```diff
@@ -146,73 +146,83 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
-            let reader = blob.as_blob_reader(buffer).await?;
-            blob_info.push((section, blob.clone(), reader, blob_size));
+            let mut reader = blob.as_blob_reader(buffer).await?;
+            // For the first section, seek to the start position
+            let initial_offset = if is_first {
+                reader.seek_to(start_byte_offset)?;
+                start_byte_offset
+            } else {
+                0
+            };
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs:207]: we should return ItemPruned error in this case I think.

[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs:207]: reverted to old behavior

File: storage/src/journal/segmented/fixed.rs [hunk 6]
```diff
@@ -423,7 +472,7 @@ mod tests {
 
             let items = {
                 let stream = journal
-                    .replay(0, NZUsize!(1024))
+                    .replay(0, 0, NZUsize!(1024))
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs:474]: Should we add some test with non-zero start?

File: storage/src/journal/segmented/fixed.rs [hunk 7]
```diff
@@ -319,6 +334,40 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+
+        Ok(())
+    }
+
+    /// Ensure a section exists, creating an empty blob if needed.
+    ///
+    /// This is used to maintain the invariant that at least one blob always exists
+    /// (the "tail" blob), which allows reconstructing journal size on reopen.
+    pub async fn ensure_section_exists(&mut self, section: u64) -> Result<(), Error> {
+        self.manager.get_or_create(section).await?;
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs:367]: I think we should just call `self.manager.get_or_create()`? It does the same thing?

File: storage/src/journal/segmented/fixed.rs [hunk 8]
```diff
@@ -319,6 +334,40 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+
+    /// Initialize a section with a specific number of zero-filled items.
+    ///
+    /// This creates the section's blob and fills it with `item_count` items worth of zeros.
+    /// The data is written through the Append wrapper which handles checksums properly.
+    ///
+    /// # Arguments
+    /// * `section` - The section number to initialize
+    /// * `item_count` - Number of zero-filled items to write
+    pub async fn init_section_at_size(
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs]: We should add a test for this?

File: storage/src/journal/segmented/fixed.rs [hunk 9]
```diff
@@ -187,76 +187,91 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
-            blob_info.push((section, replay));
+            let blob_size = blob.size().await;
+            let mut replay = blob.replay(buffer).await?;
+            // For the first section, seek to the start position
+            let initial_position = if section == start_section {
+                let start = start_position * Self::CHUNK_SIZE_U64;
+                if start > blob_size {
+                    return Err(Error::ItemOutOfRange(start_position));
+                }
+                replay.seek_to(start as usize).await?;
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs]: I think `seek_to` should take `u64`?

Comment (patrick-ogrady) [reply storage/src/journal/segmented/fixed.rs]: If I didn't follow this advice elsewhere, please lmk

File: storage/src/journal/segmented/fixed.rs [hunk 10]
```diff
@@ -187,76 +187,91 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
-            let replay = blob.replay(buffer).await?;
-            blob_info.push((section, replay));
+            let blob_size = blob.size().await;
+            let mut replay = blob.replay(buffer).await?;
+            // For the first section, seek to the start position
+            let initial_position = if section == start_section {
+                let start = start_position * Self::CHUNK_SIZE_U64;
+                if start > blob_size {
+                    return Err(Error::ItemOutOfRange(start_position));
+                }
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs:202]: I'm assuming this byte-filling is already an established pattern?

File: storage/src/journal/contiguous/fixed.rs [hunk 11]
```diff
@@ -426,226 +313,103 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+        // Calculate the tail section.
+        let tail_section = self.size / self.items_per_blob;
+
+        // Cap to tail section. The tail section is guaranteed to exist by our invariant.
+        let min_section = std::cmp::min(target_section, tail_section);
+
+        let pruned = self.inner.prune(min_section).await?;
         if pruned {
-            self.pruning_boundary = new_oldest_blob * self.cfg.items_per_blob.get();
+            self.pruning_boundary = min_section * self.items_per_blob;
```
Comment (patrick-ogrady) [inline storage/src/journal/contiguous/fixed.rs]: Do we need to maintain the pruning_boundary here? I "think" it is already tracked by `inner` (can call oldest_section())?

File: storage/src/journal/contiguous/fixed.rs [hunk 12]
```diff
@@ -106,205 +86,155 @@ pub struct Config {
     pub write_buffer: NonZeroUsize,
 }
 
-/// A contiguous journal with fixed-size entries.
+/// Implementation of `Journal` storage.
 ///
-/// Items are stored sequentially across multiple blobs, with automatic blob rotation
-/// when a blob reaches `items_per_blob` items.
-///
-/// # Repair
```
Comment (patrick-ogrady) [inline storage/src/journal/contiguous/fixed.rs:114]: imho we should retain this repair comment?

File: runtime/src/utils/buffer/pool/read.rs [hunk 13]
```diff
@@ -315,6 +315,31 @@ impl<B: Blob> Replay<B> {
...
+
+    /// Seeks to `offset` in the blob, returning `Err(BlobInsufficientLength)` if `offset` exceeds
+    /// the blob size.
+    pub async fn seek_to(&mut self, offset: usize) -> Result<(), Error> {
+        if offset as u64 > self.reader.blob_size() {
+            return Err(Error::BlobInsufficientLength);
+        }
+
+        // Clear existing buffer state
+        self.buffer.buffers.clear();
```
Comment (patrick-ogrady) [inline runtime/src/utils/buffer/pool/read.rs]: I think this would be better handled with a dedicated `.reset()` or `.clear()` method instead of introspecting?

File: storage/src/journal/contiguous/fixed.rs (hunk 12 above)
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:114]: it's not quite accurate though. If there's invalid internal segment we return Corruption error rather than try to repair it.  The only "repair" is to rewind the tail blob if it's incomplete.

File: storage/src/journal/segmented/fixed.rs (hunk 10 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs:202]: yes, it's previous behavior for state syncing that starts in the middle of the blob.

File: storage/src/journal/contiguous/fixed.rs (hunk 11 above)
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs]: no we don't, changed. Technically I guess we don't really need to maintain size() either, should I remove that caching as well?

File: storage/src/journal/contiguous/fixed.rs (hunk 12 above)
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:114]: Oh hmm I guess it is accurate since the segmented journal does this on init now (not replay).

Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:114]: OK clarified in both segmented and contiguous that that blobs are trimmed of incomplete items on init to handle the case where an underlying write fails at a page boundary and the last item in the page gets truncated.

File: storage/src/journal/segmented/fixed.rs (hunk 9 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: my bad, changed.

File: storage/src/journal/segmented/fixed.rs (hunk 7 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs:367]: this is called from contiguous::fixed journal, are you ok with this introspection?

File: runtime/src/utils/buffer/pool/read.rs (hunk 13 above)
Comment (roberto-bayardo) [reply runtime/src/utils/buffer/pool/read.rs]: sorry misunderstood the first request, fixed now.

File: storage/src/journal/segmented/fixed.rs (hunk 6 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs:474]: Done. (There was one in contiguous which exercises this, but probably good to exercise directly as well.)

File: storage/src/journal/segmented/fixed.rs (hunk 8 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: added

File: storage/src/journal/contiguous/fixed.rs [hunk 14]
```diff
@@ -106,205 +91,203 @@ pub struct Config {
...
-                .map_err(Error::Runtime)?;
-            let index = match name.try_into() {
-                Ok(index) => u64::from_be_bytes(index),
-                Err(nm) => return Err(Error::InvalidBlobName(hex(&nm))),
-            };
-            debug!(blob = index, size, "loaded blob");
-            blobs.insert(index, (blob, size));
-        }
 
-        // Check that there are no gaps in the historical blobs.
```
Comment (roberto-bayardo) [inline storage/src/journal/contiguous/fixed.rs:188]: Removed this integrity checking during init() in favor of integrity checking during replay. This means you only check integrity of data that matters (old corrupted blobs may just get quietly pruned..) and you also aren't subject to silent replay failures should corruption happen after init.

File: runtime/src/utils/buffer/pool/read.rs [hunk 15]
```diff
@@ -315,6 +323,25 @@ impl<B: Blob> Replay<B> {
...
+        }
+
+        self.buffer.clear();
+        self.exhausted = false;
+
+        let page_size = self.reader.logical_page_size as u64;
+        self.reader.blob_page = offset / page_size;
+
+        let remainder = (offset % page_size) as usize;
+        assert!(self.ensure(remainder).await?, "bounds already checked");
```
Comment (patrick-ogrady) [inline runtime/src/utils/buffer/pool/read.rs]: Do we need to call `ensure`? I think the application should be calling this before it reads?

Not to mention, I think this is ensuring all the bytes up to the start offset (which seems wasteful)?

File: storage/src/journal/contiguous/fixed.rs [hunk 16]
```diff
@@ -106,205 +91,203 @@ pub struct Config {
...
+        let pos_in_section = position % self.items_per_blob;
+        (section, pos_in_section)
     }
 
     /// Sync any pending updates to disk.
     pub async fn sync(&mut self) -> Result<(), Error> {
-        self.synced.inc();
-        debug!(blob = self.tail_index, "syncing blob");
-        self.tail.sync().await.map_err(Error::Runtime)
+        self.inner.sync_all().await
```
Comment (patrick-ogrady) [inline storage/src/journal/contiguous/fixed.rs]: Do we need to sync_all or just the tip?

File: runtime/src/utils/buffer/pool/read.rs (hunk 15 above)
Comment (roberto-bayardo) [reply runtime/src/utils/buffer/pool/read.rs]: Without it, test_replay_seek_to fails, but let me look if there's a better way.

File: storage/src/journal/segmented/fixed.rs [hunk 17]
```diff
@@ -58,18 +58,13 @@ pub struct Config {
...
 ///
-/// Like
-/// [sqlite](https://github.com/sqlite/sqlite/blob/8658a8df59f00ec8fcfea336a2a6a4b5ef79d2ee/src/wal.c#L1504-L1505)
-/// and
-/// [rocksdb](https://github.com/facebook/rocksdb/blob/0c533e61bc6d89fdf1295e8e0bcee4edb3aef401/include/rocksdb/options.h#L441-L445),
-/// the first invalid data read will be considered the new end of the journal (and the
-/// underlying [Blob] will be truncated to the last valid item). Repair occurs during
-/// init by checking each blob's size.
+/// If a write fails at a page boundary, there's a chance the journal will contain a "partial" item
+/// at the end of the last appended blob. All blobs are checked on init and any incomplete items are
```
Comment (patrick-ogrady) [inline storage/src/journal/segmented/fixed.rs]: This "incomplete" items handles things that span multiple pages correctly?

File: storage/src/journal/contiguous/fixed.rs (hunk 16 above)
Comment (patrick-ogrady) [reply storage/src/journal/contiguous/fixed.rs]: The same is also true of `variable` I suppose (may speed things up a good bit 👀 )

File: runtime/src/utils/buffer/pool/read.rs (hunk 15 above)
Comment (roberto-bayardo) [reply runtime/src/utils/buffer/pool/read.rs]: (Since we set blob_page though I don't think it's buffering anything other than the page containing the requested seek offset)

Comment (roberto-bayardo) [reply runtime/src/utils/buffer/pool/read.rs]: ok fixed, though had to introduce a `first_fill_skip` to handle the first buffer fill after seeking properly.  Extended the test case to exercise this appropriately.

File: storage/src/journal/segmented/fixed.rs (hunk 17 above)
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: It should, since the truncated size logic is simply:

`let valid_size = size - (size % Self::CHUNK_SIZE_U64);`

... which is page size agnostic.  I could add a test though?

File: storage/src/journal/segmented/fixed.rs [hunk 18]
```diff
@@ -319,6 +329,40 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+
+    /// Initialize a section with a specific number of zero-filled items.
+    ///
+    /// This creates the section's blob and fills it with `item_count` items worth of zeros.
+    /// The data is written through the Append wrapper which handles checksums properly.
+    ///
+    /// # Arguments
+    /// * `section` - The section number to initialize
+    /// * `item_count` - Number of zero-filled items to write
+    pub async fn init_section_at_size(
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: done

File: storage/src/journal/segmented/fixed.rs [hunk 19]
```diff
@@ -319,6 +329,40 @@ impl<E: Storage + Metrics, A: CodecFixedShared> Journal<E, A> {
...
+        blob.resize(target_size).await?;
+
+        Ok(())
+    }
+
+    /// Ensure a section exists, creating an empty blob if needed.
+    ///
+    /// This is used to maintain the invariant that at least one blob always exists
+    /// (the "tail" blob), which allows reconstructing journal size on reopen.
+    pub async fn ensure_section_exists(&mut self, section: u64) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/segmented/fixed.rs]: done

File: runtime/src/utils/buffer/pool/read.rs (hunk 15 above)
Comment (patrick-ogrady) [reply runtime/src/utils/buffer/pool/read.rs]: Pushed a small cleanup to this

File: storage/src/journal/contiguous/fixed.rs (hunk 14 above)
Comment (patrick-ogrady) [reply storage/src/journal/contiguous/fixed.rs:188]: This is similar to variable (and probably is ok).

**Review by patrick-ogrady — APPROVED** [review]

**Review by patrick-ogrady — APPROVED** [review]
